sdk/src/types/block/semantic/unlock.rs

```diff
@@ -13,58 +13,63 @@ impl SemanticValidationContext<'_> {
     ///
     pub fn address_unlock(&mut self, address: &Address, unlock: &Unlock) -> Result<(), TransactionFailureReason> {
         match (address, unlock) {
-            (Address::Ed25519(ed25519_address), Unlock::Signature(unlock)) => {
-                if self.unlocked_addresses.contains(address) {
-                    return Err(TransactionFailureReason::SemanticValidationFailed);
-                }
+            (Address::Ed25519(ed25519_address), unlock) => match unlock {
+                Unlock::Signature(unlock) => {
+                    if self.unlocked_addresses.contains(address) {
+                        return Err(TransactionFailureReason::DirectUnlockableAddressUnlockInvalid);
+                    }
 
-                let Signature::Ed25519(signature) = unlock.signature();
+                    let Signature::Ed25519(signature) = unlock.signature();
 
-                if signature
-                    .is_valid(self.transaction_signing_hash.as_ref(), ed25519_address)
-                    .is_err()
-                {
-                    return Err(TransactionFailureReason::UnlockSignatureInvalid);
-                }
+                    if signature
+                        .is_valid(self.transaction_signing_hash.as_ref(), ed25519_address)
+                        .is_err()
```

Comment:
hmm, weird. Does `is_valid` also return a bool we should check?

Reply:
`pub fn is_valid(&self, message: &[u8], address: &Ed25519Address) -> Result<(), Error>`

Reply:
Can fail for multiple reasons

Reply:
probably shouldn't be called `is_valid` then, unless this isn't in our lib?

Reply:
`validate` probably

Reply:
Ok I can change it